### ISSUE TITLE
PropertyAffordance section is missing `type` specific in "vocabulary terms" table

### ISSUE THREAD

**Issue by thjaeckle** [w3c/wot-thing-description#1390]
During implementation of a TM->TD generator I noticed that the section [PropertyAffordance](https://w3c.github.io/wot-thing-description/#propertyaffordance) is - when strictly comparing with the JsonSchema - missing several `type` specific terms.

It is declared to be "[...] a Subclass of the `InteractionAffordance` Class and the `DataSchema` Class."
[DataSchema](https://w3c.github.io/wot-thing-description/#dataschema) does only declare "Vocabulary terms" common for all data `type`s Subclasses.
The subclass specific terms are declared in the subclasses themselves, e.g.:
* for [ArraySchema](https://w3c.github.io/wot-thing-description/#arrayschema): `items`, `minItems`, `maxItems`
* for [ObjectSchema](https://w3c.github.io/wot-thing-description/#objectschema): `properties`, `required`
* etc.

Looking at the [InteractionAffordance](https://w3c.github.io/wot-thing-description/#interactionaffordance) class those terms are also not defined.

The JsonSchema correctly contains the definitions that those `type` specific terms are part of the `"property_element"`:
https://github.com/w3c/wot-thing-description/blob/10000c7a6b11fb2d978682f07ebbfbc07203076c/validation/td-json-schema-validation.json#L518-L572

The specification however does not.
So from my point the PropertyAffordance must either be a subclass of all of the available `DataSchema` subclasses (which would prevent redundancy in the spec) or it must declare all type specific terms in its "vocabulary terms" table.

Does this make sense?

**Comment by egekorkan** [comment]
Well PropertyAffordance is indeed a subclass of DataSchema, please see the screenshot below:
![image](https://user-images.githubusercontent.com/20195407/153730951-3f9d0b33-775b-492f-9164-b5d09d5eb844.png)

**Comment by egekorkan** [comment]
I think I understand the problem now. E.g. ObjectSchema inherits DataSchema, PropertyAffordance inherits DataSchema but there is no relationship between PropertyAffordance and ObjectSchema. 

In a way, we need to say that PropertyAffordance inherits DataSchema and all the other types?

**Comment by JKRhb** [comment]
I guess this is an old discussion point, but would it also be an option (probably not until TD 2.0) to adjust the JSON Schema to reflect the current class structure? I think it would be a bit more intuitive and would make the specification easier to implement if a property/data schema like

```json
{
  "type": "string",
  "minimum": 2,
  "minItems": 3
}
```

wasn't allowed.

In this case, however, you would probably need to have one inherited `PropertyAffordance` class for every `DataSchema` class.

**Comment by egekorkan** [comment]
Even when the inheritance/relationship is done in JSON Schema, it would still need additional ways to describe keys that are not allowed since by design, additional terms are allowed in TDs. Probably something that uses property names (http://json-schema.org/understanding-json-schema/reference/object.html#property-names) and then does not allow specific words in specific data schema children

**Comment by sebastiankb** [comment]
from today's TD call:
- we just started to study the issue and found out that the `properties` term is missing in the dataSchema definition
- at least in TD 1.0, this is showing up in the [Overview Figure 2](https://www.w3.org/TR/wot-thing-description/#overview)
- we should fix this before CR

**Comment by ektrah** [comment]
IMHO the fundamental problem is that a property affordance is not a kind of data schema. An object schema is a kind of data schema, a number schema is a kind of data schema, etc. However, a property affordance is something where you can read or write a value, and that *value* conforms to some data schema (which we want to describe in the property affordance description).

**Comment by thjaeckle** [comment]
@ektrah true - for `actions` and `events` this is clearly separated in the elements `input`/`output` (for actions) and `data` (for events). 
For TD 2.0 that would probably be a better approach for properties as well - e.g. to have a `data` element aside from the property's `title`.

**Comment by egekorkan** [comment]
> For TD 2.0 that would probably be a better approach for properties as well - e.g. to have a data element aside from the property's title.

**Warning: This opens a can of worms.**

This was actually the case before (around 2018). Below are the ways that this was done:
- https://www.w3.org/TR/2018/WD-wot-thing-description-20180405/#property via `schema`
- https://www.w3.org/TR/2017/WD-wot-thing-description-20170914/#property via `outputData`

There are also multiple discussions around this:
- Here is one comment from @mkovatsc : https://github.com/w3c/wot-thing-description/issues/66#issuecomment-365391260
- https://github.com/w3c/wot/issues/290
- https://github.com/w3c/wot/issues/393

If I were to summarize the discussion from my _memory_:
1. From developer's perspective, it was more convenient to have constructs like `mything.properties.temperature` that results in the value rather than `mything.properties.temperature.outputData`.
2. Somewhat similar to above, a property has a value and that should have a single data schema. It is weird if a property can have different data schema for writing and reading. (This is however happening in multiple IoT platforms).
3. We wanted to use `readOnly` keyword for whole property and these from JSON Schema so property as a Data Schema made somewhat sense.

My personal opinion:
- We should open a separate issue if we want this back
- I find the first argument I have summarized to be actually very weak since the TD structure is not how values are represented upon operations on the Thing
- I would like to see this as well since it helps with quite a bit of stuff

Note: I was not really active in WoT activities when these discussions took place so I would like to get opinions of @mkovatsc, @benfrancis , @draggett , @sebastiankb, and probably more people.

**Comment by ektrah** [comment]
TD 2.0 is a separate issue.

For TD 1.1, we have a syntax and I think we just need to touch up a bit the semantics. For example,

```json
{
  "title": "Current Room Temperature",
  "type": "number",
  "minimum": -273,
  "maxumum": 655.35,
  "unit": "qudt:DEG_C",
  "readOnly": true
}
```

describes a property with title "Current Room Temperature". When reading and writing this property, the property value conforms to the data schema

```json
{
  "type": "number",
  "minimum": -273,
  "maxumum": 655.35,
  "unit": "qudt:DEG_C",
  "readOnly": true
}
```

So, on the syntactic level, it the property description looks like a small extension of the data schema. On the semantic level, however, I don't think it makes a lot of sense to say that the affordance for reading the current room temperature is an instance of a data schema.

How could we describe the semantics more accurately without changing the syntax? Above, @JKRhb suggested "to have one inherited `PropertyAffordance` class for every `DataSchema` class". I think that could lead to the right direction. The inheritance tree would then look like this:

* InteractionAffordance
  * PropertyAffordance
    * PropertyAffordanceWithArrayValue
    * PropertyAffordanceWithBooleanValue
    * PropertyAffordanceWithNumberValue
    * PropertyAffordanceWithIntegerValue
    * PropertyAffordanceWithObjectValue
    * PropertyAffordanceWithStringValue
    * PropertyAffordanceWithNullValue
  * ActionAffordance
  * EventAffordance

(probably with better names). The different kinds of `PropertyAffordance` would have the same *syntax* as their JSON Schema counterpart, but the *semantics* would be different in that the fields describe the value when reading and writing the property and not the property as such.

Would that work?
